### PR TITLE
[autodiff] Support clear gradients by type

### PR DESCRIPTION
**Pull request by erizmr**
Related issue = #5731

### CODE REVIEW

File: python/taichi/_kernels.py [hunk 1]
```diff
@@ -224,7 +224,7 @@ def arr_vulkan_layout_to_field_normal_layout(vk_arr: ndarray_type.ndarray(),
 def clear_gradients(_vars: template()):
     for I in grouped(ScalarField(Expr(_vars[0]))):
         for s in static(_vars):
-            ScalarField(Expr(s))[I] = 0
+            ScalarField(Expr(s))[I] = ops.cast(0, dtype=s.get_dt())
```
Comment (ailzhang) [inline python/taichi/_kernels.py:227]: OOC: why is this necessary? :P

File: python/taichi/ad/_ad.py [hunk 2]
```diff
@@ -92,7 +95,7 @@ def grad(self):
         self.gradient_evaluated = True
 
 
-def clear_all_gradients():
+def clear_all_gradients(gradient_type=SNodeGradType.ADJOINT):
```
Comment (ailzhang) [inline python/taichi/ad/_ad.py:98]: my gut feeling from this api name is that by default it clears all types of gradients :P

**Review by ailzhang — COMMENTED** [review]
thanks!!

File: python/taichi/_kernels.py (hunk 1 above)
Comment (erizmr) [reply python/taichi/_kernels.py:227]: These are annoying `i32 -> u8 lose precision` warnings if without this.

File: python/taichi/ad/_ad.py (hunk 2 above)
Comment (erizmr) [reply python/taichi/ad/_ad.py:98]: The api name is indeed somehow misleading. I think the initial meaning of the api name might be clearing all the adjoint fields (because there is no other grad type at that time).

**Review by ailzhang — APPROVED** [review]
LGTM!
